pkg/v1/tarball/write.go

```diff
@@ -29,83 +29,112 @@ import (
-func MultiWriteToFile(p string, tagToImage map[name.Tag]v1.Image) error {
+func MultiWriteToFile(p string, tagToImage map[name.Tag]v1.Image, updates chan<- v1.Update) error {
 	refToImage := make(map[name.Reference]v1.Image, len(tagToImage))
 	for i, d := range tagToImage {
 		refToImage[i] = d
 	}
-	return MultiRefWriteToFile(p, refToImage)
+	return MultiRefWriteToFile(p, refToImage, updates)
 }
 
 // MultiRefWriteToFile writes in the compressed format to a tarball, on disk.
 // This is just syntactic sugar wrapping tarball.MultiRefWrite with a new file.
-func MultiRefWriteToFile(p string, refToImage map[name.Reference]v1.Image) error {
+func MultiRefWriteToFile(p string, refToImage map[name.Reference]v1.Image, updates chan<- v1.Update) error {
 	w, err := os.Create(p)
 	if err != nil {
 		return err
 	}
 	defer w.Close()
 
-	return MultiRefWrite(refToImage, w)
+	return MultiRefWrite(refToImage, w, updates)
 }
 
 // Write is a wrapper to write a single image and tag to a tarball.
-func Write(ref name.Reference, img v1.Image, w io.Writer) error {
-	return MultiRefWrite(map[name.Reference]v1.Image{ref: img}, w)
+func Write(ref name.Reference, img v1.Image, w io.Writer, updates chan<- v1.Update) error {
+	return MultiRefWrite(map[name.Reference]v1.Image{ref: img}, w, updates)
 }
 
 // MultiWrite writes the contents of each image to the provided reader, in the compressed format.
 // The contents are written in the following format:
 // One manifest.json file at the top level containing information about several images.
 // One file for each layer, named after the layer's SHA.
 // One file for the config blob, named after its SHA.
-func MultiWrite(tagToImage map[name.Tag]v1.Image, w io.Writer) error {
+func MultiWrite(tagToImage map[name.Tag]v1.Image, w io.Writer, updates chan<- v1.Update) error {
 	refToImage := make(map[name.Reference]v1.Image, len(tagToImage))
 	for i, d := range tagToImage {
 		refToImage[i] = d
 	}
-	return MultiRefWrite(refToImage, w)
+	return MultiRefWrite(refToImage, w, updates)
 }
 
 // MultiRefWrite writes the contents of each image to the provided reader, in the compressed format.
 // The contents are written in the following format:
 // One manifest.json file at the top level containing information about several images.
 // One file for each layer, named after the layer's SHA.
 // One file for the config blob, named after its SHA.
-func MultiRefWrite(refToImage map[name.Reference]v1.Image, w io.Writer) error {
+func MultiRefWrite(refToImage map[name.Reference]v1.Image, w io.Writer, updates chan<- v1.Update) error {
 	tf := tar.NewWriter(w)
 	defer tf.Close()
 
 	imageToTags := dedupRefToImage(refToImage)
 	var m Manifest
 
+	// add up the total sizes of everything we are downloading
+	var size, complete int64
+	var sendUpdate = func() {
+		if updates != nil {
+			updates <- v1.Update{
+				Total:    size,
+				Complete: complete,
+			}
+		}
+	}
+
+	for img, name := range imageToTags {
```

Comment:
I didn't think about this before, but we need to keep in mind how to avoid breaking streaming images. This code assumes it's cheap to calculate an image's `Manifest()`, but that's not always the case. We might need to rethink this design a bit so that it's possible for us to send updates cheaply. E.g. maybe the total size is something callers should calculate themselves?

Reply:
> assumes it's cheap to calculate an image's Manifest()

Is it not? By the time someone gets to `MultiRefWrite()`, have they not already got the manifest? Doesn't the map in the parameters to the func of `refToImage map[name.Reference]v1.Image` already contain the actual images, which contain the manifest? 

Or do we have cases where calling `img.Manifest()` actually will pull at that moment? Either way, manifest certainly is tiny cost compared with the layers themselves, and will have to be called anyways.

> maybe the total size is something callers should calculate themselves?

I would be hesitant about that. One of the beauties of this library is that it makes it so much easier to "just get images" (and manipulate them, to some degree). Making them dig in to calculate would be harder.

Reply:
> Is it not? By the time someone gets to MultiRefWrite(), have they not already got the manifest?

Not always: https://github.com/google/go-containerregistry/tree/master/pkg/v1/stream

Basically, with stream.Layer and mutating images, the manifest might be lazily calculated, which depends on the streaming layer being consumed already. I'm not sure if this should be blocking, though.

> I would be hesitant about that. One of the beauties of this library is that it makes it so much easier to "just get images" (and manipulate them, to some degree). Making them dig in to calculate would be harder.

Seems reasonable -- another way to address my concern would be to only calculate these values if the caller passes in WithProgress so that we avoid paying the cost if we can.

Reply:
> another way to address my concern would be to only calculate these values if the caller passes in WithProgress

And yet another argument in favour of your functional options. :-)

Reply:
Take a look at the latest commit. I think this addresses it.